include/camp/array.hpp

```diff
@@ -0,0 +1,291 @@
+                                              const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] != rhs[i]) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator!=(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      return !(lhs == rhs);
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator<(const array<T, N>& lhs,
+                                             const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] < rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
+      }
+
+      return false;
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator<=(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] < rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator>(const array<T, N>& lhs,
+                                             const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] > rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
+      }
+
+      return false;
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator>=(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] > rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr T& get(array<T, N>& a) noexcept {
+      return a[I];
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr T&& get(array<T, N>&& a) noexcept {
+      return move(a[I]);
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr const T& get(const array<T, N>& a) noexcept {
+      return a[I];
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr const T&& get(const array<T, N>&& a) noexcept {
+      return move(a[I]);
+   }
+
+   // TODO: Investigate the implementation from cppreference
```

Comment:
I'm having a little trouble with the implementation:

```
namespace detail {
      template <class T, std::size_t N, std::size_t... I>
      CAMP_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
         to_array_impl(T (&a)[N], idx_seq<I...>)
      {
         return {{a[I]...}};
      }

      template <class T, std::size_t N, std::size_t... I>
      CAMP_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
         to_array_impl(T (&&a)[N], idx_seq<I...>)
      {
         return {{move(a[I])...}};
      }
   }

   template <class T, std::size_t N>
   CAMP_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
      to_array(T (&a)[N])
   {
      return detail::to_array_impl(a, make_idx_seq_t<N>{});
   }

   template <class T, std::size_t N>
   CAMP_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
      to_array(T (&&a)[N])
   {
      return detail::to_array_impl(move(a), make_idx_seq_t<N>{});
   }
```

Test case:

```
int temp[3] = {1, 2, 10};
camp::array<int, 3> a = camp::to_array(temp);
camp::array<int, 3> b = camp::to_array(camp::move(temp));
```

Error messages (there seem to be issues with non-type template parameters):

```
In file included from camp/test/array.cpp:1:
camp/include/camp/array.hpp:263:14: error: no matching function for call to 'to_array_impl'
      return detail::to_array_impl(a, make_idx_seq_t<N>{});
             ^~~~~~~~~~~~~~~~~~~~~
camp/test/array.cpp:351:34: note: in instantiation of function template specialization 'camp::to_array<int, 3UL>' requested here
   camp::array<int, 3> a = camp::to_array(temp);
                                 ^
camp/include/camp/array.hpp:246:10: note: candidate template ignored: substitution failure [with T = int, N = 3]: deduced non-type template argument does not have the same type as the corresponding template parameter ('long' vs 'std::size_t' (aka 'unsigned long'))
         to_array_impl(T (&a)[N], idx_seq<I...>)
         ^
camp/include/camp/array.hpp:253:10: note: candidate template ignored: substitution failure [with T = int, N = 3]: deduced non-type template argument does not have the same type as the corresponding template parameter ('long' vs 'std::size_t' (aka 'unsigned long'))
         to_array_impl(T (&&a)[N], idx_seq<I...>)
         ^
camp/include/camp/array.hpp:270:14: error: no matching function for call to 'to_array_impl'
      return detail::to_array_impl(move(a), make_idx_seq_t<N>{});
             ^~~~~~~~~~~~~~~~~~~~~
camp/test/array.cpp:352:34: note: in instantiation of function template specialization 'camp::to_array<int, 3UL>' requested here
   camp::array<int, 3> b = camp::to_array(camp::move(temp));
                                 ^
camp/include/camp/array.hpp:246:10: note: candidate template ignored: substitution failure [with T = int, N = 3]: deduced non-type template argument does not have the same type as the corresponding template parameter ('long' vs 'std::size_t' (aka 'unsigned long'))
         to_array_impl(T (&a)[N], idx_seq<I...>)
         ^
camp/include/camp/array.hpp:253:10: note: candidate template ignored: substitution failure [with T = int, N = 3]: deduced non-type template argument does not have the same type as the corresponding template parameter ('long' vs 'std::size_t' (aka 'unsigned long'))
         to_array_impl(T (&&a)[N], idx_seq<I...>)
         ^
2 errors generated.
```

Reply:
If I use `auto N` and `auto... I` instead of `std::size_t N` and `std::size_t... I`, it compiles fine. But I think the use of `auto` for non-type template parameters isn't supported until C++17.

Reply:
I can conditionally use the implementation with auto based on whether C++17 is being used.

Reply:
I think camp has its own idx_t or something like that that should be used with make_idx_seq_t.

Reply:
But the std::array class prescribes std::size_t as the index type, so I'm not sure I want to go that route.

Reply:
Try this https://github.com/LLNL/camp/blob/79c320fa09db987923b56884afdc9f82f4b70fc4/include/camp/number.hpp#L117

Reply:
Thanks, Jason. That worked.